Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2722,6 +2728,148 @@ extension BrowserViewController: ContextMenuHelperDelegate {
+            self.customSearchEngineButton?.userInteractionEnabled = true
+        }
+
+        guard let webContentView = findWebViewFirstResponder(webView),
+            let input = webContentView.performSelector(Selector("inputAccessoryView")),
+            let inputView = input.takeUnretainedValue() as? UIInputView,
+            let nextButton = inputView.valueForKey("_nextItem") as?  UIBarButtonItem,
+            let nextButtonView = nextButton.valueForKey("view") as? UIView else {
+                //failed to find the inputView instead lets use the findInPage bar
+                addCustomSearchButtonToFindInPage()
+                return
+            }
+
+            inputView.addSubview(self.customSearchEngineButton!)
+            self.customSearchEngineButton!.snp_remakeConstraints { make in
+                make.left.equalTo(nextButtonView.snp_right).offset(20)
+                make.width.equalTo(inputView.snp_height)
+                make.top.equalTo(nextButtonView.snp_top)
+                make.height.equalTo(inputView.snp_height)
+            }
+    }
+
+    /**
+     This adds the customSearchButton to the findInPage container
+     for cases where the inputAccessoryView could not be found for example
+     on the iPad where it does not exist.
+     **/
+    func addCustomSearchButtonToFindInPage() {
+        let view = SearchAddBar()
+        if self.customSearchEngineButton == nil {
+            self.customSearchEngineButton = self.searchEngineButton()
+        }
+        view.addSubview(self.customSearchEngineButton!)
+        findInPageContainer.addSubview(view)
+
+        view.snp_makeConstraints { make in
+            make.edges.equalTo(findInPageContainer)
+            make.height.equalTo(UIConstants.ToolbarHeight)
+        }
+
+        self.customSearchEngineButton!.snp_remakeConstraints { make in
+            make.width.equalTo(view.snp_height)
+            make.height.equalTo(view.snp_height)
+            make.left.equalTo(view.snp_left).offset(20)
+            make.top.equalTo(view.snp_top)
+        }
+
+        updateViewConstraints()
+        self.findInPageContainer.layoutIfNeeded()
+    }
+
+    func searchEngineButton() -> UIButton {
+        let searchButton = UIButton()
+        searchButton.setImage(UIImage(named: "AddSearch")?.imageWithRenderingMode(.AlwaysTemplate), forState: .Normal)
+        searchButton.addTarget(self, action: #selector(BrowserViewController.addCustomSerchEngineForFocusedElement), forControlEvents: .TouchUpInside)
+        return searchButton
+    }
+
+    func addCustomSerchEngineForFocusedElement() {
+        guard let webView = tabManager.selectedTab?.webView else {
+            return
+        }
+        webView.evaluateJavaScript("__firefox__.searchQueryForField()") { (result, _) in
+            guard let searchParams = result as? [String:String] else {
+                //we failed. YOU WERE THE CHOSEN ONE. YOU WERE SUPPPOSE TO BRING BALANCE TO THE SEARCH ENGINES
```

Comment:
Heh, funny comment -- but it'd be useful to have information here on why this might fail.